File: modules/vela-wallet-webview/android/src/main/java/com/velawallet/webview/WalletWebView.kt
// WalletWebView (Android) — a wallet-owned android.webkit.WebView for the in-app
// dApp browser. Android counterpart of the iOS WKWebView view.
//
// Responsibilities (see docs/dapp-browser/ARCHITECTURE.md):
//   • Inject the shared EIP-1193/6963 provider at DOCUMENT START, main frame only
//     (AndroidX WebViewCompat.addDocumentStartJavaScript) so window.ethereum is
//     present before page JS and immune to the page CSP.
//   • Receive provider requests on the "velaBridge" WebMessageListener, stamping
//     the TRUSTED sourceOrigin + isMainFrame the framework supplies — NEVER an
//     origin from the page payload.
//   • Deliver responses/events via window.__velaRespond / window.__velaEmit.
//   • Report navigation lifecycle for the URL bar + settle-on-navigation.
package com.velawallet.webview

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.webkit.WebMessageCompat
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.events.RCTEventEmitter
import org.json.JSONArray
import org.json.JSONObject

@SuppressLint("SetJavaScriptEnabled")
class WalletWebView(context: Context) : WebView(context) {

    private var injectedSource: String? = null // provider bundle (readiness signal)
    private var documentStartSupported = false // false → late-inject fallback on old WebView
    private var pendingUrl: String? = null
    private var isLoading = false
    private var lastFavicon = "" // resolved favicon URL for the current document
    private var lastSeq = 0 // highest processed outbox seq (native <- RN deliveries)

    init {
        settings.javaScriptEnabled = true
        settings.domStorageEnabled = true
        settings.databaseEnabled = true
        settings.mediaPlaybackRequiresUserGesture = true

        webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                isLoading = true
                lastFavicon = "" // new document — drop the previous page's favicon
                // Fallback for pre-DOCUMENT_START_SCRIPT devices: inject as early as we
                // can (not before the first byte, but before most dApps read ethereum).
                if (!documentStartSupported) {
                    injectedSource?.let { view?.evaluateJavascript(it, null) }
                }
                emitNavigation(true)
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                isLoading = false
                emitNavigation(false)
                resolveFavicon(view) // async — re-emits nav once the favicon is known
            }

            // http(s) loads here; every other scheme is handed to the OS (mailto:,
            // tel:, wc:, itms-apps:, app links) or dropped (javascript:/file:/data:).
            override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                val uri = request?.url ?: return false
                val scheme = uri.scheme?.lowercase() ?: ""
                if (scheme == "http" || scheme == "https" || scheme == "about") return false
                if (scheme != "javascript" && scheme != "file" && scheme != "data") {
                    try {
                        context.startActivity(Intent(Intent.ACTION_VIEW, uri).apply {
                            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        })
                    } catch (_: Exception) { /* no handler app — swallow */ }
                }
                return true // never navigate the WebView to a non-web scheme
            }

            override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
                if (request?.isForMainFrame != true) return // ignore sub-resource failures
                isLoading = false
                emitNavigation(false, error?.description?.toString() ?: "Failed to load")
            }

            // SPA client-side routing (history.pushState/replaceState) fires this but
            // NOT onPageStarted/Finished — refresh the URL bar + canGoBack for it.
            override fun doUpdateVisitedHistory(view: WebView?, url: String?, isReload: Boolean) {
                emitNavigation(isLoading)
            }
        }

        // page -> native: the injected shim calls window.velaBridge.postMessage(string).
        if (WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER)) {
            WebViewCompat.addWebMessageListener(
                this,
                "velaBridge",
                setOf("*"),
            ) { _, message, sourceOrigin, isMainFrame, _ ->
                onBridgeMessage(message, sourceOrigin?.toString() ?: "", isMainFrame)
            }
        }
    }

    // --- props (from the view manager) --------------------------------------

    fun setInjectedJavaScriptSource(source: String) {
        if (injectedSource != null || source.isEmpty()) return
        injectedSource = source
        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            WebViewCompat.addDocumentStartJavaScript(this, source, setOf("*"))
            documentStartSupported = true
        }
        // Load regardless of feature support — an old WebView still browses (with the
        // onPageStarted late-inject fallback) instead of hanging on a blank screen.
        loadIfReady()
    }

    fun setSource(url: String) {
        pendingUrl = url
        // Defer so the injectedJavaScript prop (same commit) registers before load.
        post { loadIfReady() }
    }

    private fun loadIfReady() {
        val url = pendingUrl ?: return
        if (injectedSource == null) return // wait for the provider bundle prop
        pendingUrl = null
        loadUrl(url)
    }

    /**
     * native -> page deliveries, as a seq-tracked JSON queue prop (parity with iOS).
     * Props flow reliably under the New-Arch interop layer; process items whose seq
     * exceeds the last one we ran.
     */
    fun setOutboxSource(json: String) {
        val arr = try { JSONArray(json) } catch (e: Exception) { return }
        for (i in 0 until arr.length()) {
            val item = arr.optJSONObject(i) ?: continue
            val seq = item.optInt("seq", -1)
            if (seq <= lastSeq) continue
            lastSeq = seq
            when (item.optString("t")) {
                "res" -> evaluateJavascript(
                    "window.__velaRespond && window.__velaRespond(" +
                        "${JSONObject.quote(item.optString("id"))}, ${jsLit(item.opt("result"))}, ${jsLit(item.opt("error"))});",
                    null,
                )
                "evt" -> evaluateJavascript(
                    "window.__velaEmit && window.__velaEmit(" +
                        "${JSONObject.quote(item.optString("event"))}, ${jsLit(item.opt("data"))});",
                    null,
                )
                "nav" -> when (item.optString("action")) {
                    "back" -> if (canGoBack()) goBack()
                    "forward" -> if (canGoForward()) goForward()
                    "reload" -> reload()
                }
            }
        }
    }

    private fun jsLit(v: Any?): String = when (v) {
        null, JSONObject.NULL -> "null"
        is JSONObject -> v.toString()
        is JSONArray -> v.toString()
        is String -> JSONObject.quote(v)
        is Boolean, is Int, is Long, is Double -> v.toString()
        else -> JSONObject.quote(v.toString())
    }

    // --- native -> page ------------------------------------------------------

    fun deliverResponse(id: String, resultJson: String, errorJson: String) {
        val idLit = JSONObject.quote(id)
        val resultLit = if (resultJson.isEmpty()) "undefined" else resultJson
        val errorLit = if (errorJson.isEmpty()) "null" else errorJson
        evaluateJavascript("window.__velaRespond && window.__velaRespond($idLit, $resultLit, $errorLit);", null)
    }

    fun deliverEvent(event: String, dataJson: String) {
        val eventLit = JSONObject.quote(event)
        val dataLit = if (dataJson.isEmpty()) "null" else dataJson
        evaluateJavascript("window.__velaEmit && window.__velaEmit($eventLit, $dataLit);", null)
    }

    // --- page -> native ------------------------------------------------------

    private fun onBridgeMessage(message: WebMessageCompat, sourceOrigin: String, isMainFrame: Boolean) {
        val raw = message.data ?: return
        val obj = try { JSONObject(raw) } catch (e: Exception) { return }
        if (obj.optString("dir") != "req") return // ignore the "ready" ping etc.

        val payload: WritableMap = Arguments.createMap().apply {
            putString("requestId", obj.optString("id"))
            putString("method", obj.optString("method"))
            putArray("params", jsonArrayToWritable(obj.optJSONArray("params") ?: JSONArray()))
            putString("origin", sourceOrigin)       // TRUSTED — from the framework, not the page
            putBoolean("isMainFrame", isMainFrame)
        }
        emitEvent("onProviderRequest", payload)
    }

    private fun emitNavigation(loading: Boolean, error: String = "") {
        val payload = Arguments.createMap().apply {
            putString("url", url ?: "")
            putString("title", title ?: "")
            putBoolean("canGoBack", canGoBack())
            putBoolean("canGoForward", canGoForward())
            putBoolean("loading", loading)
            putString("error", error)
            putString("favicon", lastFavicon)
        }
        emitEvent("onNavigationChange", payload)
    }

    /** Resolve the page's declared favicon (absolute URL) and re-emit navigation. */
    private fun resolveFavicon(view: WebView?) {
        view?.evaluateJavascript(FAVICON_JS) { raw ->
            val href = try {
                if (raw.isNullOrEmpty() || raw == "null") "" else JSONArray("[$raw]").optString(0, "")
            } catch (e: Exception) {
                ""
            }
            if (href.isNotEmpty()) {
                lastFavicon = href
                emitNavigation(isLoading)
            }
        }
    }

    companion object {
        // Shared verbatim with the iOS view — reads the best-declared favicon as an
        // absolute URL (link.href resolves relatives), falling back to /favicon.ico.
        private const val FAVICON_JS =
            "(function(){try{var s=['link[rel~=\"icon\"]','link[rel=\"shortcut icon\"]'," +
                "'link[rel=\"apple-touch-icon\"]'];for(var i=0;i<s.length;i++){var l=document.querySelector(s[i]);" +
                "if(l&&l.href)return l.href;}return location.origin+'/favicon.ico';}catch(e){return '';}})()"
    }

    private fun emitEvent(name: String, payload: WritableMap) {
        val reactContext = context as ReactContext
        reactContext.getJSModule(RCTEventEmitter::class.java).receiveEvent(id, name, payload)
    }

    // --- JSON -> RN writable conversion -------------------------------------

    private fun jsonArrayToWritable(arr: JSONArray): WritableArray {
        val out = Arguments.createArray()
        for (i in 0 until arr.length()) {
            when (val v = arr.get(i)) {
                is JSONObject -> out.pushMap(jsonObjectToWritable(v))
                is JSONArray -> out.pushArray(jsonArrayToWritable(v))
                is Boolean -> out.pushBoolean(v)
                is Int -> out.pushInt(v)
                is Long -> out.pushDouble(v.toDouble())
                is Double -> out.pushDouble(v)
                JSONObject.NULL -> out.pushNull()
                else -> out.pushString(v.toString())
            }
        }
        return out
    }

    private fun jsonObjectToWritable(obj: JSONObject): WritableMap {
        val out = Arguments.createMap()
        val keys = obj.keys()
        while (keys.hasNext()) {
            val k = keys.next()
            when (val v = obj.get(k)) {
                is JSONObject -> out.putMap(k, jsonObjectToWritable(v))
                is JSONArray -> out.putArray(k, jsonArrayToWritable(v))
                is Boolean -> out.putBoolean(k, v)
                is Int -> out.putInt(k, v)
                is Long -> out.putDouble(k, v.toDouble())
                is Double -> out.putDouble(k, v)
                JSONObject.NULL -> out.putNull(k)
                else -> out.putString(k, v.toString())
            }
        }
        return out
    }
}

File: modules/vela-wallet-webview/ios/WalletWebView.swift
// WalletWebView (iOS) — a wallet-owned WKWebView for the in-app dApp browser.
//
// Responsibilities (see docs/dapp-browser/ARCHITECTURE.md):
//   • Inject the shared EIP-1193/6963 provider (inpage.js + protocol.js + shim)
//     at DOCUMENT START, main frame only, so `window.ethereum` exists before any
//     page JS runs and is immune to the page CSP.
//   • Receive provider requests on the `velaBridge` message handler and bubble
//     them to RN, stamping the TRUSTED origin from `frameInfo.securityOrigin`
//     and `frameInfo.isMainFrame` — NEVER an origin the page put in the payload.
//   • Deliver responses/events back into the page via `window.__velaRespond` /
//     `window.__velaEmit` (the injected shim owns the vela-1193 envelope).
//   • Report navigation lifecycle so RN can drive the URL bar and settle pending
//     signing on navigation.
//
// This is a classic RN view (legacy RCTViewManager), copied into the app target
// by plugins/with-native-modules.js, mirroring vela-passkey / vela-app-group.
import Foundation
import UIKit
import WebKit
import React

/// Reads the page's best-declared favicon as an ABSOLUTE url (link.href resolves
/// relative paths), falling back to the origin's /favicon.ico. Shared verbatim with
/// the Android view so both platforms report the same thing.
private let FAVICON_JS = """
(function(){try{var s=['link[rel~="icon"]','link[rel="shortcut icon"]','link[rel="apple-touch-icon"]'];\
for(var i=0;i<s.length;i++){var l=document.querySelector(s[i]);if(l&&l.href)return l.href;}\
return location.origin+'/favicon.ico';}catch(e){return '';}})()
"""

@objc(WalletWebView)
class WalletWebView: UIView, WKScriptMessageHandler, WKNavigationDelegate, WKUIDelegate {

  // MARK: RN-exported callbacks
  @objc var onProviderRequest: RCTDirectEventBlock?
  @objc var onNavigationChange: RCTDirectEventBlock?

  private var webView: WKWebView?
  private var pendingURL: URL?
  private var providerScriptInstalled = false
  private var lastSeq = 0 // highest processed outbox seq (native ← RN deliveries)
  private var lastFavicon = "" // resolved favicon URL for the current document

  // MARK: Lifecycle

  override init(frame: CGRect) {
    super.init(frame: frame)
    setUpWebView()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setUpWebView()
  }

  private func setUpWebView() {
    let config = WKWebViewConfiguration()
    config.userContentController = WKUserContentController()
    config.websiteDataStore = .default() // persistent cookies / localStorage (browser-completeness comes later)
    if #available(iOS 14.0, *) {
      config.defaultWebpagePreferences.allowsContentJavaScript = true
    }

    let wv = WKWebView(frame: bounds, configuration: config)
    // Page → native channel. Register on the LIVE controller (post-init): WKWebView
    // COPIES the configuration, and while user scripts survive that copy (injection
    // works), a script-message-handler added to the pre-init controller does NOT —
    // which is exactly why injection succeeded but messages never arrived.
    // `.postMessage` in the shim then lands in userContentController(_:didReceive:).
    //
    // Register via a WEAK proxy: WKUserContentController retains its message handler
    // strongly, and self owns the webView (→ configuration → userContentController),
    // so registering `self` directly is a retain cycle that leaks the whole WKWebView
    // on every browser open (deinit never runs). The proxy holds self weakly.
    wv.configuration.userContentController.add(WeakScriptMessageHandler(self), name: "velaBridge")
    wv.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    wv.navigationDelegate = self // weak on WKWebView — no cycle
    wv.uiDelegate = self // weak on WKWebView — for the prompt()-based page→native bridge
    wv.allowsBackForwardNavigationGestures = true
    addSubview(wv)
    webView = wv
  }

  // MARK: Props (set by the view manager)

  /// The full injected bundle (INJECTED_PROVIDER_JS). Registered as a
  /// document-start user script the first time it is set.
  @objc func setInjectedJavaScript(_ source: NSString?) {
    guard !providerScriptInstalled, let source = source as String?, !source.isEmpty else { return }
    let script = WKUserScript(source: source, injectionTime: .atDocumentStart, forMainFrameOnly: true)
    webView?.configuration.userContentController.addUserScript(script)
    providerScriptInstalled = true
    loadIfReady()
  }

  /// The URL to load. Deferred to the next run-loop tick so the injectedJavaScript
  /// prop (same commit batch) is registered before the first navigation begins.
  @objc func setSourceURL(_ value: NSString?) {
    guard let str = value as String?, let url = URL(string: str) else { return }
    pendingURL = url
    DispatchQueue.main.async { [weak self] in self?.loadIfReady() }
  }

  private func loadIfReady() {
    guard providerScriptInstalled, let url = pendingURL, let wv = webView else { return }
    pendingURL = nil
    wv.load(URLRequest(url: url))
  }

  /// native → page deliveries, as a seq-tracked JSON queue prop. Props flow
  /// reliably under the New-Arch interop layer (unlike NativeModules view-manager
  /// commands, which are unreachable in bridgeless mode). Process items whose seq
  /// exceeds the last one we ran.
  @objc func setOutbox(_ value: NSString?) {
    guard let s = value as String?, let data = s.data(using: .utf8),
          let arr = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]] else { return }
    for item in arr {
      guard let seq = item["seq"] as? Int, seq > lastSeq else { continue }
      lastSeq = seq
      switch item["t"] as? String {
      case "res":
        evaluate("window.__velaRespond && window.__velaRespond(\(jsString(item["id"] as? String ?? "")), \(jsonLit(item["result"])), \(jsonLit(item["error"])));")
      case "evt":
        evaluate("window.__velaEmit && window.__velaEmit(\(jsString(item["event"] as? String ?? "")), \(jsonLit(item["data"])));")
      case "nav":
        switch item["action"] as? String {
        case "back": goBack()
        case "forward": goForward()
        case "reload": reloadPage()
        default: break
        }
      default:
        break
      }
    }
  }

  // MARK: Native → page (invoked by the view manager commands)

  func deliverResponse(id: String, resultJson: String, errorJson: String) {
    let idLit = jsString(id)
    let resultLit = resultJson.isEmpty ? "undefined" : resultJson
    let errorLit = errorJson.isEmpty ? "null" : errorJson
    evaluate("window.__velaRespond && window.__velaRespond(\(idLit), \(resultLit), \(errorLit));")
  }

  func deliverEvent(event: String, dataJson: String) {
    let eventLit = jsString(event)
    let dataLit = dataJson.isEmpty ? "null" : dataJson
    evaluate("window.__velaEmit && window.__velaEmit(\(eventLit), \(dataLit));")
  }

  func goBack() { if webView?.canGoBack == true { webView?.goBack() } }
  func goForward() { if webView?.canGoForward == true { webView?.goForward() } }
  func reloadPage() { webView?.reload() }

  private func evaluate(_ js: String) {
    DispatchQueue.main.async { [weak self] in self?.webView?.evaluateJavaScript(js, completionHandler: nil) }
  }

  // MARK: WKScriptMessageHandler (page → native)

  func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
    // Kept as a fallback; on-device the interop-registered WKScriptMessageHandler
    // does not deliver, so the shim uses the custom-scheme channel (below) instead.
    if message.name == "velaBridge", let raw = message.body as? String {
      processBridge(raw, from: message.frameInfo)
    }
  }

  /// WKUIDelegate: the prompt()-based page→native bridge. The shim calls
  /// `window.prompt('velawvbridge:' + payload)`; we read it and dismiss silently.
  func webView(_ webView: WKWebView, runJavaScriptTextInputPanelWithPrompt prompt: String,
               defaultText: String?, initiatedByFrame frame: WKFrameInfo,
               completionHandler: @escaping (String?) -> Void) {
    if prompt.hasPrefix("velawvbridge:") {
      // SECURITY: stamp origin/isMainFrame from the INITIATING frame, so a
      // cross-origin subframe that calls window.prompt('velawvbridge:…') directly
      // (it can — prompt is a native global, no shim needed) cannot masquerade as
      // the main frame. The WebViewTransport 4100s any isMainFrame=false request.
      processBridge(String(prompt.dropFirst("velawvbridge:".count)), from: frame)
      completionHandler(nil) // dismiss without showing a dialog
      return
    }
    // Not our bridge — a real page prompt(). Return nil (= user dismissed) rather
    // than defaultText (which falsely reports "user confirmed the default value").
    completionHandler(nil)
  }

  /// Process one page→native `vela-1193` request envelope (from either channel).
  /// `frame` is the WKFrameInfo the request actually came from — its
  /// `securityOrigin` and `isMainFrame` are the TRUSTED source of both fields
  /// (never a value the page put in the payload, never the top-frame URL).
  private func processBridge(_ raw: String, from frame: WKFrameInfo) {
    guard let data = raw.data(using: .utf8),
          let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
          (obj["dir"] as? String) == "req" else { return } // ignore the "ready" ping etc.
    onProviderRequest?([
      "requestId": obj["id"] as? String ?? "",
      "method": obj["method"] as? String ?? "",
      "params": obj["params"] ?? [],
      "origin": originString(from: frame.securityOrigin),
      "isMainFrame": frame.isMainFrame,
    ])
  }

  // MARK: WKNavigationDelegate (navigation lifecycle)

  /// Reliable JS→native channel: intercept the custom bridge scheme the shim
  /// navigates a throwaway iframe to (the WKScriptMessageHandler doesn't fire
  /// under the RN New-Arch interop view).
  func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
    guard let url = navigationAction.request.url else { decisionHandler(.allow); return }
    let scheme = url.scheme?.lowercased() ?? ""

    // Reliable JS→native channel: the shim navigates a throwaway iframe to this
    // custom scheme (the WKScriptMessageHandler doesn't fire under the RN interop).
    if scheme == "velawvbridge" {
      let full = url.absoluteString
      if let r = full.range(of: "velawvbridge://m/") {
        let enc = String(full[r.upperBound...])
        // SECURITY: stamp from navigationAction.sourceFrame (the frame that issued
        // the bridge navigation), not the top frame — same guard as the prompt path.
        if let decoded = enc.removingPercentEncoding {
          processBridge(decoded, from: navigationAction.sourceFrame)
        }
      }
      decisionHandler(.cancel)
      return
    }

    // Normal web content loads in the WebView.
    if scheme == "http" || scheme == "https" || scheme == "about" || scheme == "blob" {
      decisionHandler(.allow)
      return
    }

    // External scheme (mailto:, tel:, wc:, itms-apps:, https-app-links, …): hand to
    // the OS — but ONLY for a real main-frame link tap. A programmatic redirect
    // (`location.href='facetime://…'`) or a hidden cross-origin iframe is
    // navigationType .other / a subframe, and must NOT be able to launch external
    // apps or our own velawallet:// deep link without a user gesture. Dangerous
    // local schemes (javascript:, file:, data:) are never loaded nor opened.
    decisionHandler(.cancel)
    let userTapped = navigationAction.navigationType == .linkActivated && navigationAction.sourceFrame.isMainFrame
    if userTapped, scheme != "javascript", scheme != "file", scheme != "data", scheme != "velawallet" {
      UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }
  }

  /// target=_blank / window.open — WKWebView otherwise does nothing (targetFrame
  /// is nil). Load the request in THIS view so "view tx / docs / twitter" links work.
  func webView(_ webView: WKWebView, createWebViewWith configuration: WKWebViewConfiguration,
               for navigationAction: WKNavigationAction, windowFeatures: WKWindowFeatures) -> WKWebView? {
    if navigationAction.targetFrame == nil, let url = navigationAction.request.url,
       let scheme = url.scheme?.lowercased(), scheme == "http" || scheme == "https" {
      webView.load(URLRequest(url: url))
    }
    return nil
  }

  func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
    lastFavicon = "" // new document — drop the previous page's favicon
    // Emit loading EARLY (before didCommit) so the DNS/TLS/first-byte wait — the
    // slowest phase — shows a progress bar instead of a blank screen.
    emitNavigation(loading: true)
  }
  func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
    emitNavigation(loading: true)
  }
  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    emitNavigation(loading: false)
    resolveFavicon() // async — emits a follow-up nav update once the URL is known
  }
  func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
    emitNavigation(loading: false, error: error)
  }
  func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
    emitNavigation(loading: false, error: error)
  }

  /// Resolve the page's declared favicon (absolute URL) and re-emit navigation so
  /// the URL bar can show it. Falls back to /favicon.ico.
  private func resolveFavicon() {
    guard let wv = webView else { return }
    let urlAtCall = wv.url
    wv.evaluateJavaScript(FAVICON_JS) { [weak self] result, _ in
      guard let self = self, let w = self.webView, w.url == urlAtCall else { return } // navigated away — drop the stale result
      if let href = result as? String, !href.isEmpty {
        self.lastFavicon = href
        self.emitNavigation(loading: false)
      }
    }
  }

  private func emitNavigation(loading: Bool, error: Error? = nil) {
    guard let wv = webView else { return }
    // Ignore the benign "cancelled" that fires when we redirect an external scheme
    // or start a new load over an in-flight one — it isn't a real page failure.
    let ns = error as NSError?
    let isCancel = ns?.domain == NSURLErrorDomain && ns?.code == NSURLErrorCancelled
    onNavigationChange?([
      "url": wv.url?.absoluteString ?? "",
      "title": wv.title ?? "",
      "canGoBack": wv.canGoBack,
      "canGoForward": wv.canGoForward,
      "loading": loading,
      "error": (error != nil && !isCancel) ? (ns?.localizedDescription ?? "Failed to load") : "",
      "favicon": lastFavicon,
    ])
  }

  // MARK: Helpers

  private func originString(from o: WKSecurityOrigin) -> String {
    if o.protocol.isEmpty && o.host.isEmpty { return "" }
    let scheme = o.protocol
    let host = o.host
    let isDefaultPort = (scheme == "https" && o.port == 443) || (scheme == "http" && o.port == 80) || o.port == 0
    return isDefaultPort ? "\(scheme)://\(host)" : "\(scheme)://\(host):\(o.port)"
  }

  /// Serialize an arbitrary JSON value into a JS literal (`"0x1"`, `[…]`, `{…}`,
  /// `1`, `true`, or `null`). Wrap in an array so scalars serialize, then strip.
  private func jsonLit(_ value: Any?) -> String {
    guard let value = value, !(value is NSNull) else { return "null" }
    if let d = try? JSONSerialization.data(withJSONObject: [value]),
       let s = String(data: d, encoding: .utf8) {
      return String(s.dropFirst().dropLast())
    }
    return "null"
  }

  /// JSON-encode a string into a safe JS string literal.
  private func jsString(_ s: String) -> String {
    if let d = try? JSONSerialization.data(withJSONObject: [s]),
       let arr = String(data: d, encoding: .utf8) {
      // arr is `["..."]` — strip the surrounding brackets to get the bare literal.
      return String(arr.dropFirst().dropLast())
    }
    return "\"\""
  }

  deinit {
    webView?.configuration.userContentController.removeScriptMessageHandler(forName: "velaBridge")
    webView?.stopLoading()
  }
}

/// A weak forwarder so WKUserContentController (which retains its message handler
/// strongly) does not retain the WalletWebView — otherwise self → webView →
/// configuration → userContentController → self is a cycle that leaks the whole
/// WKWebView on every browser open. See setUpWebView().
private final class WeakScriptMessageHandler: NSObject, WKScriptMessageHandler {
  private weak var delegate: WKScriptMessageHandler?
  init(_ delegate: WKScriptMessageHandler) { self.delegate = delegate }
  func userContentController(_ controller: WKUserContentController, didReceive message: WKScriptMessage) {
    delegate?.userContentController(controller, didReceive: message)
  }
}
